### ISSUE TITLE
Incorrect type inferance after a conditional raise

### ISSUE THREAD

**Issue by macdjord** [PyCQA/pylint#8331]
### Bug description

`test.py`:
```python
"""Test"""
import typing as _t

NumericType = _t.Union[int, float, complex]
InstantiatableType = _t.Union[NumericType, str, set, None]


def instantiate(type_: str) -> InstantiatableType:
    """Instantiate the requested type"""

    if type_ == "int":
        return 1
    if type_ == "float":
        return 1.0
    if type_ == "complex":
        return 1.0j
    if type_ == "str":
        return "a"
    if type_ == "set":
        return set()

    return None


def negate(type_: str) -> NumericType:
    """Negate a value of the requested type"""

    operand: InstantiatableType = instantiate(type_)

    if not isinstance(operand, (int, float, complex)):
        raise ValueError(f"operand not a number")

    return -operand
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:33:11: E1130: bad operand type for unary -: str (invalid-unary-operand-type)
test.py:33:11: E1130: bad operand type for unary -: set (invalid-unary-operand-type)
test.py:33:11: E1130: bad operand type for unary -: NoneType (invalid-unary-operand-type)

------------------------------------------------------------------
Your code has been rated at 2.50/10 (previous run: 2.50/10, +0.00)
```

### Expected behavior

Pylint should identify that, at line 33, the type of `operand` can only be `int`, `float`, or `complex`, all of which support unary minus.

### Pylint version

```shell
pylint 2.16.2
astroid 2.14.2
Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]
```

### OS / Environment

Windows 10
